### PR TITLE
Improve resilience against memory attacks

### PR DESCRIPTION
**Pull request by droidmonkey**
[TIP]:  # ( Provide a general summary of your changes in the title above ^^ )

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ Refactor (significant modification to existing code)

## Description and Context
[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Describe the context of your change. Explain large code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" as necessary )
To reduce residual fragments of secret data in memory after
deallocation, this patch replaces the global delete operator with a
version that zeros out previously allocated memory. It makes use of
the new C++14 sized deallocation, but provides an unsized fallback
with platform-specific size deductions.

This change is only a minor mitigation and cannot protect against
buffer reallocations by the operating system or non-C++ libraries.
Thus, we still cannot guarantee all memory to be wiped after free.

As a further improvement, this patch uses libgcrypt and libsodium
to write long-lived master key component hashes into a secure
memory area and wipe it afterwards.

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )
Testing as a major part of 2.4.2

## Checklist:
[NOTE]: # ( Please go over all the following points. )
[NOTE]: # ( Again, remove any lines which don't apply. )
[NOTE]: # ( Pull Requests that don't fulfill all [REQUIRED] requisites are likely )
[NOTE]: # ( to be sent back to you for correction or will be rejected.  )
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- ✅ My change requires a change to the documentation, and I have updated it accordingly.
- ✅ I have added tests to cover my changes.

### CODE REVIEW

File: src/browser/CMakeLists.txt [hunk 1]
```diff
@@ -16,9 +16,10 @@
 
 if(WITH_XC_BROWSER)
     include_directories(${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_BINARY_DIR})
-    find_package(sodium 1.0.12 REQUIRED)
+    include_directories(SYSTEM ${sodium_INCLUDE_DIR})
```
Comment (droidmonkey) [inline src/browser/CMakeLists.txt]: Shouldn't these be ```target_include_directories``` ??

File: src/keys/PasswordKey.h [hunk 2]
```diff
@@ -30,13 +30,16 @@ class PasswordKey : public Key
...
     QByteArray rawKey() const override;
     void setPassword(const QString& password);
 
     static QSharedPointer<PasswordKey> fromRawKey(const QByteArray& rawKey);
 
 private:
-    QByteArray m_key;
+    static constexpr int SHA256_SIZE = 32;
+
+    char* m_key = nullptr;
```
Comment (droidmonkey) [inline src/keys/PasswordKey.h]: Shouldn't m_key be defined in ```Key```? It is used in all keys.

File: src/browser/CMakeLists.txt (hunk 1 above)
Comment (phoerious) [reply src/browser/CMakeLists.txt]: I don't think so. We want to use libsodium in all of KeePassXC, not only the main target.

File: src/keys/PasswordKey.h (hunk 2 above)
Comment (phoerious) [reply src/keys/PasswordKey.h]: Yes and no. I kinda want to keep it as an implementation detail. Allocation and deallocation should be in the same class and it wouldn't make a lot of sense to move it to the parent class for any other reason. Also, other keys might just be a wrapper for an external key facility that don't need an internal buffer at all.

File: src/browser/CMakeLists.txt (hunk 1 above)
Comment (droidmonkey) [reply src/browser/CMakeLists.txt]: Then this should be in the main CMakeLists.txt not conditionally added

Comment (phoerious) [reply src/browser/CMakeLists.txt]: Yeah, now that you're saying it, I don't know why I didn't move it out of the if condition.

Comment (phoerious) [reply src/browser/CMakeLists.txt]: Alright, I did add it to the global CMakeLists.txt. This one's just redundant and I removed it.

**Comment by droidmonkey** [comment]
Looks good to me. Do we need to add "Alloc.cpp" to all the intermediate libraries? You added it to Browser but I would think it would come as part of linking against Core.

**Comment by phoerious** [comment]
It should be enough to build it into the main library as it's applied during link time. I added it to browser, because it didn't link against the core lib, but the core lib is linking against browser, of course. So it's actually redundant here. I did add it to the proxy, though, since that one's a standalone binary that doesn't link against core.

**Comment by sizeofvoid** [comment]
Hallo keepasscx team,

and thanks for the next great release 2.4.2. I'm a little confused by this feature. How can Alloc.cpp work by simply adding to the build? I would be very pleased to receive a short answer.

RS

**Pull request merged**